### PR TITLE
DynamoDB sort key optimization

### PR DESCRIPTION
**Pull request by brfrn169**
In this PR, we use `KeyBytesEncoder` for the sort key to optimize Scan operations.

After this change, we no longer need local indexes for multiple clustering keys. We just need to have one primary key for clustering keys (We still need global secondary indexes for Scalar DB secondary indexes).

With `KeyBytesEncoder`, we can encode multiple clustering keys to a single byte array while preserving the sort order. And in the case of Scan operations, we encode the start clustering key and the end clustering key to byte arrays with `KeyBytesEncoder` and use them for conditions (`BETWEEN`, `>`, `>=`, `<`, `<=`) in queries. We also use `KeyBytesEncoder` for partition keys for efficiency. 

Please take a look!

### CODE REVIEW

File: core/src/main/java/com/scalar/db/storage/dynamo/DynamoAdmin.java [hunk 1]
```diff
@@ -203,73 +202,148 @@ public void createNamespace(String namespace, Map<String, String> options) {
...
-      }
-    }
-    requestBuilder.attributeDefinitions(columnsToAttributeDefinitions);
+    checkMetadata(metadata);
 
-    // build keys
-    buildPrimaryKey(requestBuilder, metadata);
-
-    // build local indexes that corresponding to clustering keys
-    buildLocalIndexes(namespace, table, requestBuilder, metadata);
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/storage/dynamo/DynamoAdmin.java:227]: We no longer need the local secondary index for clustering keys.

File: core/src/main/java/com/scalar/db/storage/dynamo/DynamoAdmin.java [hunk 2]
```diff
@@ -203,73 +202,148 @@ public void createNamespace(String namespace, Map<String, String> options) {
...
+  private void buildPrimaryKey(CreateTableRequest.Builder requestBuilder, TableMetadata metadata) {
+    List<KeySchemaElement> keySchemaElementList = new ArrayList<>();
+    keySchemaElementList.add(
+        KeySchemaElement.builder().attributeName(PARTITION_KEY).keyType(KeyType.HASH).build());
+    if (!metadata.getClusteringKeyNames().isEmpty()) {
+      keySchemaElementList.add(
+          KeySchemaElement.builder().attributeName(CLUSTERING_KEY).keyType(KeyType.RANGE).build());
+    }
+    requestBuilder.keySchema(keySchemaElementList);
+  }
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/storage/dynamo/DynamoAdmin.java:306]: We always need only one primary key even if multi-clustering key defined.

File: core/src/main/java/com/scalar/db/storage/dynamo/DynamoAdmin.java [hunk 3]
```diff
@@ -203,73 +202,148 @@ public void createNamespace(String namespace, Map<String, String> options) {
...
-
-    // build secondary indexes
     long ru = Long.parseLong(options.getOrDefault(REQUEST_UNIT, DEFAULT_REQUEST_UNIT));
-    buildGlobalIndexes(namespace, table, requestBuilder, metadata, ru);
 
-    // ru
+    CreateTableRequest.Builder requestBuilder = CreateTableRequest.builder();
+    buildAttributeDefinitions(requestBuilder, metadata);
+    buildPrimaryKey(requestBuilder, metadata);
+    buildSecondaryIndexes(namespace, table, requestBuilder, metadata, ru);
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/storage/dynamo/DynamoAdmin.java:212]: We still need global secondary indexes for Scalar DB secondary indexes.

File: core/src/main/java/com/scalar/db/storage/dynamo/DynamoOperation.java [hunk 4]
```diff
@@ -64,37 +64,31 @@ public String getGlobalIndexName(String column) {
...
-    Map<String, Value<?>> keyMap = new HashMap<>();
-    operation.getClusteringKey().ifPresent(k -> k.get().forEach(v -> keyMap.put(v.getName(), v)));
-
-    ConcatenationVisitor visitor = new ConcatenationVisitor();
-    metadata.getClusteringKeyNames().forEach(name -> keyMap.get(name).accept(visitor));
-
-    return Optional.of(visitor.build());
+    return Optional.of(
+        new KeyBytesEncoder()
+            .encode(operation.getClusteringKey().get(), metadata.getClusteringOrders()));
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/storage/dynamo/DynamoOperation.java:91]: We use `KeyBytesEncoder` to encode the concatenated clustering key.

File: core/src/main/java/com/scalar/db/storage/dynamo/DynamoOperation.java [hunk 5]
```diff
@@ -64,37 +64,31 @@ public String getGlobalIndexName(String column) {
...
-  String getConcatenatedPartitionKey() {
-    Map<String, Value<?>> keyMap = new HashMap<>();
-    operation.getPartitionKey().get().forEach(v -> keyMap.put(v.getName(), v));
-
-    ConcatenationVisitor visitor = new ConcatenationVisitor();
-    metadata.getPartitionKeyNames().forEach(name -> keyMap.get(name).accept(visitor));
-
-    return visitor.build();
+  ByteBuffer getConcatenatedPartitionKey() {
+    return new KeyBytesEncoder().encode(operation.getPartitionKey());
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/storage/dynamo/DynamoOperation.java:82]: We also use `KeyBytesEncoder` to encode the concatenated partition key for efficiency.

File: core/src/main/java/com/scalar/db/storage/dynamo/ItemSorter.java [hunk 6]
```diff
@@ -1,92 +0,0 @@
-package com.scalar.db.storage.dynamo;
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/storage/dynamo/ItemSorter.java:1]: We no longer need the realtime sort.

File: core/src/main/java/com/scalar/db/storage/dynamo/SelectStatementHandler.java [hunk 7]
```diff
@@ -1,24 +1,30 @@
 package com.scalar.db.storage.dynamo;
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/storage/dynamo/SelectStatementHandler.java:1]: I think the changes in `SelectStatementHandler` are most significant and a little bit tricky.

**Review by feeblefakie — COMMENTED** [review]
Great work! Thank you!
I left some questions mainly for a better understanding and very minor suggestions for the naming.

**Comment by brfrn169** [comment]
@feeblefakie @thongdk8 I fixed the logic of `SelectStatementHandler`. Please take a look at it again! Thanks.

**Review by feeblefakie — APPROVED** [review]
Great work! LGTM. Thank you!

**Review by thongdk8 — APPROVED** [review]
Great work! Thank you!

**Pull request merged**